doc/PROPOSAL.md

```diff
@@ -0,0 +1,163 @@
+integration with `vere`". Limiting Phase 2 to this is desirable because it i.)
+keeps the grant's scope of work at a reasonable level, and ii.) it allows more
+time for the following three ancillary runtime projects/considerations to
+achieve resolution:
+
+1. Asynchronous reads in `vere`
+2. Merging/releasing event log truncation
+   [urbit/urbit#7501](https://github.com/urbit/urbit/pull/5701)
+3. A decision on whether or not to add another process to the overall Urbit
+   runtime architecture
+
+Each of these three workstreams need to be resolved before Seguro can be
+integrated with `vere`.
+
+### Event Fragmentation Scheme
+
+FoundationDB supports a maximum value size of 100KB and recommends limiting
+value size to only 10KB for best performance. Since our data model requires
+arbitrarily sized byte buffer values for events, an event fragmentation
+algorithm is required to support events with larger payloads.
+
+Events which exceed the size limit of a single database value must be fragmented
+so that such events are stored across multiple database key/value pairs. To
+achieve event fragmentation, key/value pairs will be formatted like so:
+
+#### Keys/Values
+
+|           | encoding                                                                           | format                         |
+|-----------|------------------------------------------------------------------------------------|--------------------------------|
+| **key**   | `0x00` &#x7c; 64-bit event id (big-endian) &#x7c; 32-bit fragment id (big-endian)  | `0x00112233445566778899AABBCC` |
+| **value** | raw bytes starting at index `(i * <optimal fragment size>)` for fragment w/ id `i` | `0xdeadbeef...`                |
+
+Notes:
+- `|` = concatenation
+- The event ID and fragment ID are both contiguous, non-negative integers
+- The optimal fragment size, as documented above, will almost certainly be
+  `10 KB` = `10,000 bytes`
+- The first fragment (fragment `0`) will have a slightly different encoding for
+  the value, as documented below
+
+#### Fragment 0
+
+The value of the first fragment of an event will be split into two parts that
+are concatenated: the "header" and the "payload".
+
+#### Fragment 0 Header
+
+The fragment header is used to record the total number of fragments for the
+event. This serves two functions: it allows the read algorithm to pre-allocate
+an appropriate amount of memory on the heap for an event after reading the
+first batch out of FDB, and `memcpy` data from FDB as it arrives in batches; it
+allows the read algorithm to verify that the number of fragments returned by FDB
+is the number of fragments expected.
+
+The fragment header is a raw byte array with the following format:
+
+**128 fragments or less**
+
+| byte | meaning                | format          | example     |
+|------|------------------------|-----------------|-------------|
+| 0    | *additional* fragments | `<# fragments>` | `0111 1111` |
+
+**129 fragments or less**
```

Comment:
Shouldn't this be "129 fragments or more"?

Reply:
Yes, that's a typo